File: python/hackerrank/arrays/equalizeArray.py

from collections import Counter

def equalizeArray(arr):
    """
    Finds the minimum number of elements to delete to make all elements in the array equal.
    
    The function identifies the most frequent element in the array and calculates
    the number of deletions needed by subtracting the frequency of that element 
    from the total number of elements.

    Args:
        arr (list[int]): An array of integers.

    Returns:
        int: The minimum number of deletions required.
    """
    if not arr: return 0
    counts = Counter(arr)
    return len(arr) - counts[max(counts, key=counts.get)]

def main():
    """
    Main entry point for the script.
    
    Reads the array size and elements from standard input, validates the input length,
    and prints the result of `equalizeArray`.
    """
    n = int(input().strip())
    arr = list(map(int, input().rstrip().split()))

    if n != len(arr):
        raise ValueError(
            f"Invalid input: expected {n} elements, got {len(arr)}"
        )

    print(equalizeArray(arr))

if __name__ == '__main__':
    main()

File: golang/hackerrank/arrays/equalizeArray.go
package arrays

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

// EqualizeArray finds the minimum number of elements to delete to make all elements equal.
func EqualizeArray(arr []int) int {
	if len(arr) == 0 {
		return 0
	}
	counts := make(map[int]int)
	maxFreq := 0
	for _, v := range arr {
		counts[v]++
		if counts[v] > maxFreq {
			maxFreq = counts[v]
		}
	}
	return len(arr) - maxFreq
}

func main() {
	reader := bufio.NewReader(os.Stdin)

	line1, _ := reader.ReadString('\n')
	n, _ := strconv.Atoi(strings.TrimSpace(line1))

	line2, _ := reader.ReadString('\n')
	fields := strings.Fields(line2)
	arr := make([]int, n)
	for i := 0; i < n && i < len(fields); i++ {
		arr[i], _ = strconv.Atoi(fields[i])
	}

	fmt.Println(EqualizeArray(arr))
}
